quadriga/src/main/java/edu/asu/diging/quadriga/web/AddCollectionController.java

```diff
@@ -19,19 +25,34 @@
     @Autowired
     private CollectionManager collectionManager;
 
+    @Autowired
+    private CitesphereConnector citesphereConnector;
+
     @RequestMapping(value = "/auth/collections/add", method = RequestMethod.GET)
     public String get(Model model) {
         model.addAttribute("collectionForm", new CollectionForm());
         return "auth/addCollection";
     }
 
     @RequestMapping(value = "/auth/collections/add", method = RequestMethod.POST)
-    public String add(@Valid CollectionForm collectionForm, BindingResult result, RedirectAttributes redirectAttrs) {
+    public String add(Model model, @Valid CollectionForm collectionForm, BindingResult result, RedirectAttributes redirectAttrs) {
         if (result.hasErrors()) {
             return "auth/addCollection";
         }
-
-        collectionManager.addCollection(collectionForm.getName(), collectionForm.getDescription());
+        
+        List<CitesphereAppInfo> citesphereApps = citesphereConnector.getCitesphereApps();
+        HashSet<String> appSet = new HashSet<String>(
+                citesphereApps.stream().map(app -> app.getClientId()).collect(Collectors.toList()));
+        
+        for (String clientId : collectionForm.getApps()) {
+            if (!appSet.contains(clientId)) {
+                model.addAttribute("collectionForm", collectionForm);
+                return "auth/addCollection";
+            }
+        }
```

Comment:
This seems like business logic (checking that the app id is valid) and therefore does not belong in a controller. A hint is that you have the same code in the edit controller ;) 

Also, you haven't removed the problem of what happens if citesphere doesn't respond or respond slowly.  Now Quadriga would hang when the user clicks on save. I'm not sure what the best approach is. Maybe cache the results and use the cached data for this check here? It wouldn't guarantee that the apps are still valid but we could deal with the problem of deleted apps somewhere else.